service-config/src/main/java/com/palantir/config/service/ServiceDiscoveryConfiguration.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.config.service;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.palantir.remoting.ssl.SslConfiguration;
+import com.palantir.tokens.auth.BearerToken;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.immutables.value.Value.Immutable;
+import org.immutables.value.Value.Lazy;
+import org.immutables.value.Value.Style;
+
+/**
+ * Configuration class that contains a map of {@code serviceName}s and their respective {@link ServiceConfiguration}s
+ * and provides utility methods for apps to build service clients.
+ */
+@Immutable
+@JsonDeserialize(as = ImmutableServiceDiscoveryConfiguration.class)
+@Style(get = "*", visibility = Style.ImplementationVisibility.PACKAGE)
+public abstract class ServiceDiscoveryConfiguration {
```

Comment:
Because of how we customize the discovery block in https://github.palantir.build/elements/foundry/blob/develop/foundry-build/var/conf/foundry-build.yml#L41 I need a way to construct a SDC programmatically, like a .of(): https://github.palantir.build/elements/foundry/blob/develop/foundry-services/src/main/java/com/palantir/foundry/services/config/ServiceDiscoveryConfiguration.java#L24